akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala

```diff
@@ -0,0 +1,339 @@
+    watch(recoveredFsmRef)
+    recoveredFsmRef ! SubscribeTransitionCallBack(testActor)
+
+    expectMsg(CurrentState(recoveredFsmRef, Shopping))
+
+    within(0.9 seconds, 1.1 seconds) {
+      expectMsg(Transition(recoveredFsmRef, Shopping, Inactive))
+    }
+
+    expectNoMsg(0.9 seconds)    //randomly chosen delay, less than the timeout, before stopping the FSM
+    recoveredFsmRef ! PoisonPill
+    expectTerminated(recoveredFsmRef)
+
+    recoveredFsmRef = createPersistentFSMActor(persistenceId, dummyReportActorRef)
+    watch(recoveredFsmRef)
+    recoveredFsmRef ! SubscribeTransitionCallBack(testActor)
+
+    expectMsg(CurrentState(recoveredFsmRef, Inactive))
+
+    within(1.9 seconds, 2.1 seconds) {
+      expectTerminated(recoveredFsmRef)
+    }
+  }
+
+  /**
+   * Factory method for creating Persistent FSM Actor under test
+   * This method makes the test reusable for different Persistent FSM implementations,
+   * as long as they implement the contract in exactly the same way as [[akka.persistence.PersistentFSMActorSpec.SamplePersistentFsm]]
+   */
+  def createPersistentFSMActor(persistenceId: String, reportActor: ActorRef): ActorRef =
+    system.actorOf(Props(new SamplePersistentFsm(persistenceId, reportActor)))
+
+}
+
+/**
+ * [[akka.persistence.AbstractPersistentFsmActor]] spec, based on [[akka.persistence.PersistentFSMActorSpec]]
+ * [[akka.persistence.SampleJavaPersistentFsm]] has to implement the same contract as [[akka.persistence.PersistentFSMActorSpec.SamplePersistentFsm]]
+ */
+class AbstractPersistentFSMSpec extends PersistentFSMActorSpec {
+  override def createPersistentFSMActor(persistenceId: String, reportActor: ActorRef): ActorRef =
+    system.actorOf(Props(new SampleJavaPersistentFsm(persistenceId, reportActor)))
+}
+
+object PersistentFSMActorSpec {
+  //State name
+  sealed trait UserState extends FSMState
+  case object LookingAround extends UserState {
+    override def identifier: String = "Looking Around"
+  }
+  case object Shopping extends UserState {
+    override def identifier: String = "Shopping"
+  }
+  case object Inactive extends UserState {
+    override def identifier: String = "Inactive"
+  }
+  case object Paid extends UserState {
+    override def identifier: String = "Paid"
+  }
+
+  case class Item(id: String, name: String, price: Float)
+
+  //State data
+  sealed trait ShoppingCart {
+    def addItem(item: Item): ShoppingCart
+    def empty(): ShoppingCart
+  }
+  case object EmptyShoppingCart extends ShoppingCart {
+    def addItem(item: Item) = NonEmptyShoppingCart(item::Nil)
+    def empty() = this
+  }
+  case class NonEmptyShoppingCart(items: Seq[Item]) extends ShoppingCart {
+    def addItem(item: Item) = NonEmptyShoppingCart(items :+ item)
+    def empty() = EmptyShoppingCart
+  }
+
+  //Commands
+  sealed trait Command
+  case class AddItem(item: Item) extends Command
+  case object Buy extends Command
```

Comment:
ok, I missed that
later there should be a doc sample, and I was thinking this is that one, and then it should be implemented in scala and java standalone